[MTS/TL/HH][111418676] Fetch data using data attrs

michaelstalker opened the pull request:
https://www.pivotaltracker.com/story/show/111418676

**This PR looks very large. The code changes to the source files only constitute a small portion of this PR, however. Most changes are to the compiled code in the `/dist` folder, or to the tests.**

We need to find an element's item, section (which we call the subgroup
inside the jquery-autotagging.js codebase), and value. This code change
is based off of Rent.com's JavaScript tracking code.

We may need to implement a more nuanced way of finding the clicked
element; the Rent.com code deals with the target's parent element when
deciding what the click target is.

michaelstalker committed:
[MTS/TL/HH][111418676] Fetch data using data attrs

https://www.pivotaltracker.com/story/show/111418676

We need to find an element's item, section (which we call the subgroup
inside the jquery-autotagging.js codebase), and value. This code change
is based off of Rent.com's JavaScript tracking code.

We may need to implement a more nuanced way of finding the clicked
element; the Rent.com code deals with the target's parent element when
deciding what the click target is.

warrenv commented on src/data_with_data_attributes.coffee:
```diff
@@ -0,0 +1,24 @@
+      isSelect(node) || node.nodeName is 'INPUT' || node.nodeName is 'TEXTAREA'
+
+    value: (node) ->
+      if isInput(node)
```
What about using 'text = if' here. Then 'text' could be removed in each of the condition blocks.

michaelstalker replied:
:+1:

michaelstalker replied:
I'll replace the temporary variable with a call to a [query function](http://refactoring.com/catalog/replaceTempWithQuery.html).

warrenv commented on src/jquery.autotagging.coffee:
```diff
@@ -24,22 +36,24 @@ define [
+      @path                = "#{document.location.pathname}#{document.location.search}"
+      @warehouseURL        = opts.warehouseURL
       @followHref = if opts.followHref? then opts.followHref else true
+      useDataTags          = opts.useDataTags || false # opts.itemDataAttribute && opts.sectionDataAttribute
```
Is the comment still needed?

michaelstalker replied:
Nope! I'll remove it.

warrenv commented on src/select_data_without_data_attributes.coffee:
```diff
@@ -0,0 +1,21 @@
+    value = $elem.find(':selected').val() or ''
+    textFormatter.replaceDoubleByteChars(value)
+
+  value: ($elem) ->
```
Can this and the code starting at line 11 be refactored? It looks like they both do the same thing.

michaelstalker replied:
:+1:

warrenv commented:
There are several tests that seem like duplicates on first look. Not sure if we want to clean those up or not. For example:

``` bash
spec/javascripts/click_data_without_data_attributes_spec.js
46:    it('returns an empty string if there is no subgroup', function () {

spec/javascripts/data_with_data_attributes_spec.js
35:    it('returns an empty string if there is no subgroup', function () {

spec/javascripts/select_data_without_data_attributes_spec.js
77:    it('returns an empty string if there is no subgroup', function () {

```

michaelstalker committed:
[MTS][111418676] Replace temp with query

I extracted a query function out of data_with_data_attributes.coffee
file. This should make the logic a little easier to follow. In the same
file, I removed the empty string used as a default value for the text.
The function that produces the text will always return a string, so we
don't need a default value.

Since two methods did the same thing in
select_data_without_data_attributes.coffee, I made one invoke the other.
We should keep both functions around, though, to preserve a consistent
data-finding interface.

michaelstalker commented:
Thanks for reviewing this! I updated the code. My preference would be to maintain the duplicated tests for an empty string value. The functions in the three files return the same value, but only because I think, there's a business rule for each saying this should happen. The default value for one could change independently of the others. What are your thoughts?

warrenv commented:
:+1:

HansHauge commented:
:+1:

Teapane commented:
:+1:
